### PR TITLE
feat: Handle SMTP auth better

### PR DESCRIPTION
**Pull request by simar7**
Fixes a case where user/pass are not specified for a vanilla SMTP email send #241 

Adds a variety of test cases via EmailSender interface.

Signed-off-by: Simar <simar@linux.com>

### CODE REVIEW

File: outputs/email.go [hunk 1]
```diff
@@ -13,10 +13,40 @@ import (
...
+	conf EmailConfig
+	send func(string, smtp.Auth, string, []string, []byte) error
+}
+
+func (e *emailSender) Send(address string, to []string, body []byte) error {
+	var auth smtp.Auth
+	if len(e.conf.Password) > 0 && len(e.conf.Username) > 0 {
+		auth = smtp.PlainAuth("", e.conf.Username, e.conf.Password, e.conf.ServerHost)
+	}
+	return e.send(address, auth, e.conf.SenderAddr, to, body)
```
Comment (simar7) [inline outputs/email.go:43]: @afdesk - this logic should handle the case where user passes empty user/pass and expects to not use an mx server.

File: outputs/email.go [hunk 2]
```diff
@@ -13,10 +13,40 @@ import (
 	"github.com/aquasecurity/postee/v2/layout"
 )
 
+type EmailSender interface {
+	Send(address string, to []string, body []byte) error
+}
+
+type EmailConfig struct {
```
Comment (AndreyLevchenko) [inline outputs/email.go:20]: I don't see why properties of Email Output can't be used. If it required for testing then
Just defining `send`function  and then override it (and keep original code) would be enough. Duplicated properties require extra efforts on modification.

File: outputs/email.go [hunk 3]
```diff
@@ -13,10 +13,40 @@ import (
 	"github.com/aquasecurity/postee/v2/layout"
 )
 
+type EmailSender interface {
```
Comment (AndreyLevchenko) [inline outputs/email.go:16]: name is same as `emailSender` defined below. It doesn't improve readability of code.

File: outputs/email.go (hunk 1 above)
Comment (AndreyLevchenko) [reply outputs/email.go:43]: method surprisingly has a logic to handle authorization and this isn't reflected in method name. Not clear why it couldn't be part of output Send method. Affects readability as well

File: outputs/email_test.go [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+	from string
+	to   []string
+	msg  []byte
+}
+
+func TestEmailOutput_Send(t *testing.T) {
+	testCases := []struct {
+		name               string
+		lookupMXFunc       func(name string) ([]*net.MX, error)
+		emailConfig        *EmailOutput
```
Comment (AndreyLevchenko) [inline outputs/email_test.go:35]: property should be named `emailOutput` (after it's type)

File: outputs/email_test.go [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+			expectedSentEmails: 0,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			var emailsSent int
+			f, r := mockSend(tc.sendError, &emailsSent)
+
+			var ec EmailOutput
```
Comment (AndreyLevchenko) [inline outputs/email_test.go:129]: Please rename variable to reflect it's type

**Review by AndreyLevchenko — CHANGES_REQUESTED** [review]

File: outputs/email_test.go (hunk 5 above)
Comment (simar7) [reply outputs/email_test.go]: fixed

File: outputs/email_test.go (hunk 4 above)
Comment (simar7) [reply outputs/email_test.go]: fixed

File: outputs/email.go (hunk 3 above)
Comment (simar7) [reply outputs/email.go]: fixed, it's been simplified as per the suggestion.

File: outputs/email.go (hunk 1 above)
Comment (simar7) [reply outputs/email.go]: I agree - this is a good point. I've simplified the code to just be able to override sendFunc without requiring additional interfaces.

File: outputs/email.go (hunk 2 above)
Comment (simar7) [reply outputs/email.go]: You're right, I've simplified it. thanks!

**Review by AndreyLevchenko — APPROVED** [review]

**Pull request merged**